pkg/fqdn/lookup.go

```diff
@@ -0,0 +1,209 @@
+	// for the same (Qname, Qtype, Qclass)
+	clientUDP, clientTCP *dns.Client
+)
+
+// ConfigFromResolvConf parses the configuration in /etc/resolv.conf and sets
+// the configuration for pkg/fqdn.
+// nameservers and opt timeout are supported.
+// search and ndots are NOT supported.
+// This call is not thread safe.
+func ConfigFromResolvConf() error {
+	dnsConf, err := dns.ClientConfigFromFile("/etc/resolv.conf")
+	if err != nil {
+		return err
+	}
+	SetDNSConfig(dnsConf)
+
+	return nil
+}
+
+// SetDNSConfig store conf in pkg/fqdn as the global config. It also creates
+// the global UDP and TCP clients used for DNS lookups in
+// DNSLookupDefaultResolver.
+// Only .Servers and .Timeout are utilized from conf.
+// This call is not thread safe.
+func SetDNSConfig(conf *dns.ClientConfig) {
+	dnsConfig = conf
+
+	clientUDP = &dns.Client{
+		Net:            "udp",
+		Timeout:        time.Duration(dnsConfig.Timeout) * time.Second,
+		SingleInflight: true,
+	}
+
+	clientTCP = &dns.Client{
+		Net:            "tcp",
+		Timeout:        time.Duration(dnsConfig.Timeout) * time.Second,
+		SingleInflight: true,
+	}
+}
+
+// DNSLookupDefaultResolver sequentially and synchronously runs a DNS lookup
+// for every name in dnsNames. It does not use net.DefaultResolver, but
+// dnsConfig from this package (populated from resolv.conf). The configured
+// servers are always queried in the same order, only moving to the next on
+// errors (such as timeouts). The names are queried in random order (as the map
+// iteration is random) but, for each name, A records and then AAAA records are
+// requested in that order.
+// It will return:
+// DNSIPs: a map of DNS names to their IPs and associated smallest TTL (only
+// contains successful lookups). CNAME records in the response are collapsed
+// into the IPs later in the response data. The CNAME TTL is considered when
+// finding the smallest TTL.
+// DNSErrors: a map of DNS names to lookup errors.
+//
+// DNSLookupDefaultResolver is used by DNSPoller when no alternative LookupDNSNames is provided.
+func DNSLookupDefaultResolver(dnsNames []string) (DNSIPs map[string]*DNSIPRecords, DNSErrors map[string]error) {
+	return doResolverLogic(lookup, dnsNames)
+}
+
+// doResolverLogic exists to allow testing the more complex logic around
+// collecting A and AAAA records, handling CNAMEs and trying different servers.
+func doResolverLogic(lookupFunc func(string, string, uint16) (*dns.Msg, error), dnsNames []string) (DNSIPs map[string]*DNSIPRecords, DNSErrors map[string]error) {
+	DNSIPs = make(map[string]*DNSIPRecords)
+	DNSErrors = make(map[string]error)
+
+	// This is the top-level list of names to query
+	for _, dnsName := range dnsNames {
+		responseData := &DNSIPRecords{TTL: math.MaxInt32}
+
+		// Query for A & AAAA for each dnsName
+	perTypeToQuery:
+		for _, dnsType := range []uint16{dns.TypeA, dns.TypeAAAA} { // the dns library doesn't use dns.Type
```

Comment:
Why not make this its own slice, `dnsTypes`?

Reply:
I switched to using the dns package's Type type. Introducing a type just to create this slice seems more likely to create confusion than not. Would it be apparent that it's just a slice of things? I would expect custom collection types to represent some complex thing vs `[]uint16`. Still, if you still find it suboptimal I can put in `dnsTypes`.